### PR TITLE
getting-started/devflowapp: initial commit

### PR DESCRIPTION
**Pull request by alexamies**

### CODE REVIEW

File: getting-started/basicapp/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,8 @@
+FROM golang:1.10
```
Comment (tbpg) [inline getting-started/basicapp/Dockerfile]: Use 1.11 (just released)?

File: getting-started/basicapp/README.md [hunk 2]
```diff
@@ -0,0 +1,362 @@
+# Building, Testing, and Optimizing Code on the Google Cloud
+This example demonstrates a web app that sends messages from a user to her or
+his friends. It demonstrates the use of Google [Cloud
```
Comment (tbpg) [inline getting-started/basicapp/README.md]: s/her or his/their/

File: getting-started/basicapp/README.md [hunk 3]
```diff
@@ -0,0 +1,362 @@
...
+Go code with an application that can be run locally, in unit tests, in 
+[App Engine Flex](https://cloud.google.com/appengine/docs/flexible/), a plain
+Docker container, and in Kubernetes. The app is instrumented for 
+[Stackdriver Profiler](https://cloud.google.com/profiler/) and 
+[Stackdriver Trace](https://cloud.google.com/trace/).
+
+
+## Project Setup
+The prerequisites for running the example locally are 
+[Go](https://golang.org/dl/), the
```
Comment (tbpg) [inline getting-started/basicapp/README.md]: Make this a bulleted list?

File: getting-started/basicapp/README.md [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+[Stackdriver Profiler](https://cloud.google.com/profiler/) and 
+[Stackdriver Trace](https://cloud.google.com/trace/).
+
+
+## Project Setup
+The prerequisites for running the example locally are 
+[Go](https://golang.org/dl/), the 
+[gcloud](https://cloud.google.com/sdk/gcloud/) command line tool, and Docker
+community edition. If you prefer not to install gcloud and Docker locally you
+can use the [Cloud Shell](https://cloud.google.com/shell/docs/) instead.
```
Comment (tbpg) [inline getting-started/basicapp/README.md]: "can use Google Cloud Shell"

?

File: getting-started/basicapp/README.md [hunk 5]
```diff
@@ -0,0 +1,362 @@
...
+## Project Setup
+The prerequisites for running the example locally are 
+[Go](https://golang.org/dl/), the 
+[gcloud](https://cloud.google.com/sdk/gcloud/) command line tool, and Docker
+community edition. If you prefer not to install gcloud and Docker locally you
+can use the [Cloud Shell](https://cloud.google.com/shell/docs/) instead. 
+
+To run this example, you will need to create a Google Cloud project with billing
+enabled. You will also need to enable the Cloud SQL API, Cloud SQL Admin API,
+the Stackdriver Profiling API, and the Stackdriver Trace API.
```
Comment (tbpg) [inline getting-started/basicapp/README.md]: Might be good to link to a cloud.google.com page with how to do these, plus direct links to enable the APIs.

File: getting-started/basicapp/README.md [hunk 6]
```diff
@@ -0,0 +1,362 @@
...
+
+If you are trying this out in your own workspace you can get the code with the
+git clone command
```
Comment (tbpg) [inline getting-started/basicapp/README.md]: Use `go get`? That way, they only get what they need and the result is in the usual place (GOPATH).

File: getting-started/basicapp/README.md [hunk 7]
```diff
@@ -0,0 +1,362 @@
...
+```
+
+If you want to try changing code then you may want to save your code in a
+private [Google Cloud Source
+repository](https://cloud.google.com/source-repositories/docs/) with the
+commands
+```
+gcloud source repos create basicapp
+git add .
+git commit -m "My private copy"
```
Comment (tbpg) [inline getting-started/basicapp/README.md]: Does this need a `git remote add` command?

File: getting-started/basicapp/README.md [hunk 8]
```diff
@@ -0,0 +1,362 @@
...
+the Stackdriver Profiling API, and the Stackdriver Trace API.
+
+If you are trying this out in your own workspace you can get the code with the
+git clone command
+```
+git clone https://github.com/GoogleCloudPlatform/golang-samples.git
+cd getting-started/basicapp
+```
+
+If you want to try changing code then you may want to save your code in a
```
Comment (tbpg) [inline getting-started/basicapp/README.md]: New heading here? Normally I think of using a GitHub fork, but private forks cost money.

File: getting-started/basicapp/README.md [hunk 9]
```diff
@@ -0,0 +1,362 @@
...
+repository](https://cloud.google.com/source-repositories/docs/) with the
+commands
+```
+gcloud source repos create basicapp
+git add .
+git commit -m "My private copy"
+```
+
+## Local development, mocks, and unit testing
+You can run locally and unit test using a mock of the messaging service.
```
Comment (tbpg) [inline getting-started/basicapp/README.md]: What messaging service?

File: getting-started/basicapp/README.md [hunk 10]
```diff
@@ -0,0 +1,362 @@
...
+
+### Running locally
+To run the app locally commands export an environment variable for use of mocks
+instead of a real database:
```
Comment (tbpg) [inline getting-started/basicapp/README.md]: This sentence was slightly hard to parse at first. Maybe something like, "To use a mock database when running the app locally, run:"?

File: getting-started/basicapp/README.md [hunk 11]
```diff
@@ -0,0 +1,362 @@
...
+go get -d -v ./...
+go run basicwebapp.go
+```
+
+Check that the application successfully responds to a HTTP request:
+```
+curl -I http://localhost:8080
+```
+
+You can now use curl to send a message to your friend a message like
```
Comment (tbpg) [inline getting-started/basicapp/README.md]: Delete "a message like"?

File: getting-started/basicapp/README.md [hunk 12]
```diff
@@ -0,0 +1,362 @@
...
+the environment variable for use of mocks with the command
+```
+unset MESSAGE_SERVICE
+```
+
+## Unit tests
+To run unit tests use the two commands, for each of the main and services
+packages:
+```
+go test
```
Comment (tbpg) [inline getting-started/basicapp/README.md]: Use `go test -v ./...`?

File: getting-started/basicapp/README.md [hunk 13]
```diff
@@ -0,0 +1,362 @@
...
+go test -v github.com/GoogleCloudPlatform/golang-samples/getting-started/basicapp/services
+```
+
+Check that the tests pass. The Go
+[httptest](https://golang.org/pkg/net/http/httptest/) package is used to
+simulate HTTP requests and writers in some of the unit tests.
+
+Run the build and execute unit tests using Cloud Build
```
Comment (tbpg) [inline getting-started/basicapp/README.md]: "To use Cloud Build, ..."? That way, it's clear this is optional and one way to build the code.

File: getting-started/basicapp/README.md [hunk 14]
```diff
@@ -0,0 +1,362 @@
...
+```
+gcloud builds submit --config build/cb-unittest.yaml .
+```
+
+View the build and test results in the cloud console under the Cloud Build menu.
```
Comment (tbpg) [inline getting-started/basicapp/README.md]: Link?

File: getting-started/basicapp/README.md [hunk 15]
```diff
@@ -0,0 +1,362 @@
...
+
+## Database
+### Setup
+This section is based on the instructions at [Using Cloud SQL with
+Go](https://cloud.google.com/go/getting-started/using-cloud-sql). Use the
+commands below from the Cloud Shell. Make sure that the Cloud SQL API and the
+Cloud SQL Admin API are enabled.
```
Comment (tbpg) [inline getting-started/basicapp/README.md]: Links to enable APIs?

File: getting-started/basicapp/README.md [hunk 16]
```diff
@@ -0,0 +1,362 @@
...
+gcloud sql users set-password root % --instance basicapp --password [PASSWORD]  
+```
+
+To see the connection details for your instance use the command
+```
+gcloud sql instances describe $INSTANCE_NAME
+```
+
+Look for the ```connectionName```, which as the form 
+```$PROJECT:$REGION:$INSTANCE_NAME```. Also, look for the IP address of the
```
Comment (tbpg) [inline getting-started/basicapp/README.md]: Newline after code fence.

File: getting-started/basicapp/README.md [hunk 17]
```diff
@@ -0,0 +1,362 @@
...
+instance.
+
+Connect to the your Cloud SQL instance with the command in the cloud shell
```
Comment (tbpg) [inline getting-started/basicapp/README.md]: Capitalize Cloud Shell.

File: getting-started/basicapp/README.md [hunk 18]
```diff
@@ -0,0 +1,362 @@
...
+```
+INSTANCE_NAME=basicapp
+gcloud sql connect $INSTANCE_NAME --user=root
+```
+
+Execute the statements in [data/database_setup.sql](data/database_setup.sql).
+
+### Working with the Database in a Local Development Environment (Optional)
+
+To install the Cloud SQL proxy locally follow the steps in [Using Cloud SQL
```
Comment (tbpg) [inline getting-started/basicapp/README.md]: Delete "the steps in"?

File: getting-started/basicapp/README.md [hunk 19]
```diff
@@ -0,0 +1,362 @@
...
+
+```
+curl -o cloud_sql_proxy https://dl.google.com/cloudsql/cloud_sql_proxy.darwin.386
+chmod +x cloud_sql_proxy
+```
+
+Start the proxy with the commands
+
+```
+PROJECT=[your project]
```
Comment (tbpg) [inline getting-started/basicapp/README.md]: I wonder if this should be `GOOGLE_CLOUD_PROJECT` to be consistent with client libs?

File: getting-started/basicapp/README.md [hunk 20]
```diff
@@ -0,0 +1,362 @@
...
+how to configure the app to connect to Cloud SQL below.
+
+Check that the application is running OK use the same curl commands as above.
+
+Stop the app with the command
+```
+docker stop basicwebapp
+```
+
+Build the Docker container with cloud build:
```
Comment (tbpg) [inline getting-started/basicapp/README.md]: Capitalize Cloud Build.

File: getting-started/basicapp/README.md [hunk 21]
```diff
@@ -0,0 +1,362 @@
...
+```
+gcloud builds submit --config build/cb-docker.yaml .
+```
+
+Run in Docker, test that it works, and stop it
+```
+gcloud auth configure-docker
+PROJECT_ID=[your project]
```
Comment (tbpg) [inline getting-started/basicapp/README.md]: Same GOOGLE_CLOUD_PROJECT comment.

File: getting-started/basicapp/README.md [hunk 22]
```diff
@@ -0,0 +1,362 @@
...
+docker run -itd --rm -p 8080:8080 \
+  --name basicwebapp \
+  --env MESSAGE_SERVICE=mock \
+  gcr.io/$PROJECT_ID/basicwebapp
+docker ps
+curl -I http://localhost:8080
+docker stop basicwebapp
+```
+
+## Deploy the app to Flex
```
Comment (tbpg) [inline getting-started/basicapp/README.md]: AppEngine Flexible

Comment (tbpg) [inline getting-started/basicapp/README.md]: Can this run on Standard?

File: getting-started/basicapp/README.md [hunk 23]
```diff
@@ -0,0 +1,362 @@
...
+
+Check that the app is being served by App Engine with the command
+```
+gcloud app browse
+```
+
+To run the app together with the Cloud SQL database follow steps as in [Using 
+Cloud SQL for 
+MySQL](https://cloud.google.com/appengine/docs/flexible/go/using-cloud-sql) for
+Flex. You can do this by editing the app.yaml file, filling in the text for the
```
Comment (tbpg) [inline getting-started/basicapp/README.md]: file then filling

File: getting-started/basicapp/README.md [hunk 24]
```diff
@@ -0,0 +1,362 @@
...
+Revert the values back to their original symbols by editing app.yaml. To deploy
+with Cloud Build, set the strings as environment variables using the command
+```
+DB_PASSWORD=[user db password]
+gcloud builds submit \
+  --substitutions=_DB_PASSWORD=$DB_PASSWORD \
+  --config build/cb-deploy.yaml .
+```
+
+The [build/cb-deploy.yaml](build/cb-deploy.yaml) file contains a [custom build
```
Comment (tbpg) [inline getting-started/basicapp/README.md]: Looks like the code fence above messed up formatting for the rest of the file - this is showing up italicized and code blocks are reversed.

File: getting-started/basicapp/README.md [hunk 25]
```diff
@@ -0,0 +1,362 @@
...
+curl -I $PROJECT_ID.appspot.com
+```
+
+## Testing and Performance Optimization
+The next activities in your development lifecycle probably include integration
+and load testing, profiling, and tracing.
+
+### Integration and Load Testing
+
+Our integration test uses curl to make sure that the application is properly
```
Comment (tbpg) [inline getting-started/basicapp/README.md]: Should "curl" be capitalized or have code blocks? Not sure what the convention is. (Applies to the whole file.)

File: getting-started/basicapp/README.md [hunk 26]
```diff
@@ -0,0 +1,362 @@
...
+configured to talk to the database and test cases for sending and retrieving
+messages are ok. 
+
+```
+gcloud builds submit --config build/cb-e2etest.yaml .
+```
+
+### Profiling
+Now that we have generate some load on the app, we can look at the performance
```
Comment (tbpg) [inline getting-started/basicapp/README.md]: generated

File: getting-started/basicapp/README.md [hunk 27]
```diff
@@ -0,0 +1,362 @@
...
+profile. You can profile the code with [Stackdriver
+Profiler](https://cloud.google.com/profiler/docs/) to identify areas
+that can be optimized. The steps here are adapted from the page [Profiling Go
```
Comment (tbpg) [inline getting-started/basicapp/README.md]: Delete "the page."

File: getting-started/basicapp/README.md [hunk 28]
```diff
@@ -0,0 +1,362 @@
...
+### Tracing
+
+The application has been instrumented for trace data collection with
+[OpenCensus](https://opencensus.io/), with export [Stackdriver Cloud
+Trace](https://opencensus.io/)for analysis and viewing of trace data.
+The method here was adapted from the [OpenCensus Stackdriver
+Codelab](https://opencensus.io/codelabs/stackdriver/#0). To see the trace data
+navigate to the Trace menu in the Cloud Console. Because of the sparse sampling
+of trace data, you many need to run a few iterations of the end-to-end to
+generate sufficient data to be displayed.
```
Comment (tbpg) [inline getting-started/basicapp/README.md]: Mention `AlwaysSample`?

File: getting-started/basicapp/README.md [hunk 29]
```diff
@@ -0,0 +1,362 @@
...
+
+## Deploy to a Kubernetes cluster
```
Comment (tbpg) [inline getting-started/basicapp/README.md]: Mention GKE?

File: getting-started/basicapp/README.md [hunk 30]
```diff
@@ -0,0 +1,362 @@
+# Building, Testing, and Optimizing Code on the Google Cloud
```
Comment (tbpg) [inline getting-started/basicapp/README.md]: Would you mind removing references to "basic"? This integrates several different products and might not be straightforward for a newcomers. :)

Comment (alexamies) [reply getting-started/basicapp/README.md]: Changed the app name to devflowapp and removed references to 'basic'. Also removed a number of parts that made the example more complex, including kubernetes, Trace, and Profiler integrations.

File: getting-started/basicapp/README.md (hunk 29 above)
Comment (alexamies) [reply getting-started/basicapp/README.md]: Added mention

File: getting-started/basicapp/README.md (hunk 28 above)
Comment (alexamies) [reply getting-started/basicapp/README.md]: Added mention

File: getting-started/basicapp/README.md (hunk 5 above)
Comment (alexamies) [reply getting-started/basicapp/README.md]: Added

File: getting-started/basicapp/README.md (hunk 27 above)
Comment (alexamies) [reply getting-started/basicapp/README.md]: changed

File: getting-started/basicapp/README.md (hunk 26 above)
Comment (alexamies) [reply getting-started/basicapp/README.md]: changed

File: getting-started/basicapp/README.md (hunk 4 above)
Comment (alexamies) [reply getting-started/basicapp/README.md]: Changed

File: getting-started/basicapp/README.md (hunk 3 above)
Comment (alexamies) [reply getting-started/basicapp/README.md]: Changed

File: getting-started/basicapp/README.md (hunk 2 above)
Comment (alexamies) [reply getting-started/basicapp/README.md]: Changed

File: getting-started/basicapp/Dockerfile (hunk 1 above)
Comment (alexamies) [reply getting-started/basicapp/Dockerfile]: Changed

File: getting-started/basicapp/README.md (hunk 6 above)
Comment (alexamies) [reply getting-started/basicapp/README.md]: Removed this section

File: getting-started/basicapp/README.md (hunk 7 above)
Comment (alexamies) [reply getting-started/basicapp/README.md]: Removed this section

File: getting-started/basicapp/README.md (hunk 8 above)
Comment (alexamies) [reply getting-started/basicapp/README.md]: Removed section

File: getting-started/basicapp/README.md (hunk 15 above)
Comment (alexamies) [reply getting-started/basicapp/README.md]: Consolidated links to enable APIs at the top

File: getting-started/basicapp/README.md (hunk 19 above)
Comment (alexamies) [reply getting-started/basicapp/README.md]: Changed to PROJECT_ID to be consistent with Cloud Build and removed GOOGLE_CLOUD_PROJECT elsewhere

File: getting-started/basicapp/README.md (hunk 21 above)
Comment (alexamies) [reply getting-started/basicapp/README.md]: Removed the section containing GOOGLE_CLOUD_PROJECT

File: getting-started/basicapp/README.md (hunk 22 above)
Comment (alexamies) [reply getting-started/basicapp/README.md]: No, the code needs to be changed to add an import to Standard.

File: getting-started/basicapp/README.md (hunk 25 above)
Comment (alexamies) [reply getting-started/basicapp/README.md]: Changed to capital (Curl).

File: getting-started/basicapp/README.md (hunk 24 above)
Comment (alexamies) [reply getting-started/basicapp/README.md]: I am not sure how to fix it, it looks ok in my editor.

File: getting-started/basicapp/README.md (hunk 23 above)
Comment (alexamies) [reply getting-started/basicapp/README.md]: Changed

File: getting-started/basicapp/README.md (hunk 22 above)
Comment (alexamies) [reply getting-started/basicapp/README.md]: Changed

File: getting-started/basicapp/README.md (hunk 20 above)
Comment (alexamies) [reply getting-started/basicapp/README.md]: Changed

File: getting-started/basicapp/README.md (hunk 18 above)
Comment (alexamies) [reply getting-started/basicapp/README.md]: changed

File: getting-started/basicapp/README.md (hunk 17 above)
Comment (alexamies) [reply getting-started/basicapp/README.md]: Changed

File: getting-started/basicapp/README.md (hunk 16 above)
Comment (alexamies) [reply getting-started/basicapp/README.md]: Added

File: getting-started/basicapp/README.md (hunk 14 above)
Comment (alexamies) [reply getting-started/basicapp/README.md]: Added link

File: getting-started/basicapp/README.md (hunk 13 above)
Comment (alexamies) [reply getting-started/basicapp/README.md]: Cahnged

File: getting-started/basicapp/README.md (hunk 10 above)
Comment (alexamies) [reply getting-started/basicapp/README.md]: Changed as suggested

File: getting-started/basicapp/README.md (hunk 9 above)
Comment (alexamies) [reply getting-started/basicapp/README.md]: Changed to MessageService interface
